File: sponge-jython/examples/script/py/knowledge_base_callbacks.py
"""
Sponge Knowledge base
Using knowledge base callbacks.
"""

from java.util.concurrent.atomic import AtomicBoolean, AtomicInteger
from org.openksavi.sponge.examples.util import TestStatus

class ReloadTrigger(Trigger):
    def onConfigure(self):
        self.withEvent("reload")
    def onRun(self, event):
        self.logger.debug("Received event: {}", event.name)
        sponge.reload()

def onInit():
    # Variables for assertions only
    sponge.setVariable("onInitCalled", AtomicBoolean(False))
    sponge.setVariable("onBeforeLoadCalled", AtomicInteger(0))
    sponge.setVariable("onLoadCalled", AtomicInteger(0))
    sponge.setVariable("onAfterLoadCalled", AtomicInteger(0))
    sponge.setVariable("onStartupCalled", AtomicBoolean(False))
    sponge.setVariable("onBeforeReloadCalled", AtomicBoolean(False))
    sponge.setVariable("onAfterReloadCalled", AtomicBoolean(False))

    sponge.logger.debug("onInit")
    sponge.getVariable("onInitCalled").set(True)

def onBeforeLoad():
    sponge.logger.debug("onBeforeLoad")
    sponge.getVariable("onBeforeLoadCalled").incrementAndGet()

def onLoad():
    sponge.logger.debug("onLoad")
    sponge.getVariable("onLoadCalled").incrementAndGet()

def onAfterLoad():
    sponge.logger.debug("onAfterLoad")
    sponge.getVariable("onAfterLoadCalled").incrementAndGet()

def onStartup():
    sponge.logger.debug("onStartup")
    sponge.getVariable("onStartupCalled").set(True)
    sponge.event("reload").sendAfter(1000)

def onShutdown():
    sponge.logger.debug("onShutdown")
    # Using Java static field because all variables will be lost after shutdown.
    TestStatus.onShutdownCalled = True

def onBeforeReload():
    sponge.logger.debug("onBeforeReload")
    sponge.getVariable("onBeforeReloadCalled").set(True)

def onAfterReload():
    sponge.logger.debug("onAfterReload")
    sponge.getVariable("onAfterReloadCalled").set(True)

File: sponge-kotlin/src/test/kotlin/examples/KnowledgeBaseCallbacks.kt
/*
 * Sponge Knowledge base
 * Using knowledge base callbacks.
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.event.Event
import org.openksavi.sponge.kotlin.KKnowledgeBase
import org.openksavi.sponge.kotlin.KTrigger
import org.openksavi.sponge.test.util.TestStatus
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class KnowledgeBaseCallbacks : KKnowledgeBase() {

    class ReloadTrigger : KTrigger() {
        override fun onConfigure() {
            withEvent("reload")
        }
        override fun onRun(event: Event) {
            logger.debug("Received event: {}", event.name)
            sponge.reload()
        }
    }

    override fun onInit() {
        // Variables for assertions only
        sponge.setVariable("onInitCalled", AtomicBoolean(false))
        sponge.setVariable("onBeforeLoadCalled", AtomicInteger(0))
        sponge.setVariable("onLoadCalled", AtomicInteger(0))
        sponge.setVariable("onAfterLoadCalled", AtomicInteger(0))
        sponge.setVariable("onStartupCalled", AtomicBoolean(false))
        sponge.setVariable("onBeforeReloadCalled", AtomicBoolean(false))
        sponge.setVariable("onAfterReloadCalled", AtomicBoolean(false))

        logger.debug("onInit")
        sponge.getVariable<AtomicBoolean>("onInitCalled").set(true)
    }

    override fun onBeforeLoad() {
        logger.debug("onBeforeLoad")
        sponge.getVariable<AtomicInteger>("onBeforeLoadCalled").incrementAndGet()
    }

    override fun onLoad() {
        logger.debug("onLoad")
        sponge.getVariable<AtomicInteger>("onLoadCalled").incrementAndGet()
    }

    override fun onAfterLoad() {
        logger.debug("onAfterLoad")
        sponge.getVariable<AtomicInteger>("onAfterLoadCalled").incrementAndGet()
    }

    override fun onStartup() {
        logger.debug("onStartup")
        sponge.getVariable<AtomicBoolean>("onStartupCalled").set(true)
        sponge.event("reload").sendAfter(1000)
    }

    override fun onShutdown() {
        logger.debug("onShutdown")
        // Using Java static field because all variables will be lost after shutdown.
        TestStatus.onShutdownCalled = true
    }

    override fun onBeforeReload() {
        logger.debug("onBeforeReload")
        sponge.getVariable<AtomicBoolean>("onBeforeReloadCalled").set(true)
    }

    override fun onAfterReload() {
        logger.debug("onAfterReload")
        sponge.getVariable<AtomicBoolean>("onAfterReloadCalled").set(true)
    }
}
